lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/cache/EvictableLoadingCache.java

```diff
@@ -0,0 +1,369 @@
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class EvictableLoadingCache<K, V>
+        extends AbstractLoadingCache<K, V>
+{
+    public static <K, V> LoadingCache<K, V> build(
+            OptionalLong expiresAfterWriteMillis,
+            OptionalLong refreshMillis,
+            long maximumSize,
+            boolean recordStats,
+            CacheLoader<K, V> cacheLoader)
+    {
+        requireNonNull(cacheLoader, "cacheLoader is null");
+
+        CacheBuilder<Object, Object> tokenCache = CacheBuilder.newBuilder();
+        CacheBuilder<Object, Object> dataCache = CacheBuilder.newBuilder();
+        if (expiresAfterWriteMillis.isPresent()) {
+            tokenCache.expireAfterWrite(expiresAfterWriteMillis.getAsLong(), MILLISECONDS);
+            dataCache.expireAfterWrite(expiresAfterWriteMillis.getAsLong(), MILLISECONDS);
+        }
+
+        if (refreshMillis.isPresent() && (expiresAfterWriteMillis.isEmpty() || expiresAfterWriteMillis.getAsLong() > refreshMillis.getAsLong())) {
+            dataCache.refreshAfterWrite(refreshMillis.getAsLong(), MILLISECONDS);
+        }
+
+        tokenCache.maximumSize(maximumSize);
+        dataCache.maximumSize(maximumSize);
+
+        if (recordStats) {
+            dataCache.recordStats();
+        }
+
+        return new EvictableLoadingCache<>(
+                EvictableCache.buildWith(tokenCache),
+                dataCache.build(new TokenClassLoader<>(cacheLoader)));
+    }
+
+    private final EvictableCache<K, Token<K>> tokensCache;
+    private final LoadingCache<Token<K>, V> dataCache;
+
+    private EvictableLoadingCache(EvictableCache<K, Token<K>> tokensCache, LoadingCache<Token<K>, V> dataCache)
+    {
+        this.tokensCache = tokensCache;
+        this.dataCache = dataCache;
+    }
+
+    @Override
+    public V get(K key)
+            throws ExecutionException
+    {
+        Token<K> token = tokensCache.get(key, () -> new Token<>(key));
+        return dataCache.get(token);
+    }
+
+    @Override
+    public ImmutableMap<K, V> getAll(Iterable<? extends K> keys)
+            throws ExecutionException
+    {
+        BiMap<K, Token<K>> keyToToken = HashBiMap.create();
```

Comment:
I would expect `ImmutableBiMap` to be less performant, since it tries to create a compacted version at the `build()` time.
It's a method local state, so immutability doesn't buy me anything